File: 19_Remove Nth Node From End of List.java
// Time complexity: O(n)
// Space complexity: O(1)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode vHead = new ListNode(0, head);

        ListNode fast = vHead, slow = vHead;

        // 1. shift fast n steps ahead
        for (int i = 0; i < n; i++) fast = fast.next;

        // 2. shift together until tail
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }

        // 3. delete node
        slow.next = slow.next.next;

        return vHead.next;
    }
}

File: 19_Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Key Insight: fast/slow ptr w/ n gaps
        fast = slow = vHead = ListNode(0, head)
        
        # 1. shift the fast ptr n nodes
        for _ in range(n): fast = fast.next

        # 2. shift the slow/fast ptr together until fast.next reached the end
        while fast.next:
            fast = fast.next
            slow = slow.next
        
        slow.next = slow.next.next

        return vHead.next